de.fu_berlin.inf.dpp.whiteboard/frontend/src/services/javabridge.js

```diff
@@ -0,0 +1,145 @@
+      let records = sxeMessage.records || [];
+      if (sxeMessage.messageType === SXE.MESSAGE_TYPES.STATE) {
+        //we ignore root element, all html whiteboard objects will be added to root element
+        records.shift();
+      }
+      records.forEach(record => this.executeRecord(record));
+      this.canvas.renderAll();
+    };
+    //sets the state of the whiteboard at the begining of the session using data from the Java Model
+    window.setState = (state) => {
+      //convert recursive data structure to linear structure
+      let records = flatten(state);
+      //ignore root element
+      records.shift();
+      //reset
+      this.viewManager.clear();
+      records.forEach(record => this.createNew(record));
+      this.canvas.renderAll();
+    };
+  }
+
+  /**
+   * executes a given record
+   * @param {Object} record
+   */
+  executeRecord(record) {
+    let action = record.type.toLowerCase();
+    let props = record.valuePairs;
+    switch (action) {
+      case SXE.ACTIONS.NEW:
+        return this.createNew(props);
+      case SXE.ACTIONS.SET:
+        return this.set(props);
+      //the current java implementation of the SXE protocol uses "SET":visible=false instead of "REMOVE"
+      //because its faster (on java level), that why it is not possible to receive a "REMOVE" action
+      case SXE.ACTIONS.REMOVE:
+        return;
+    }
+  }
+
+  /**
+   * executes a "NEW" record
+   * @param {Object} recordProps
+   */
+  createNew(recordProps) {
+    // the value HAS to be false to ignore creation
+    // if the value is undefined or not given, proceed normally
+    if (recordProps.visible === false)
+      return;
+
+    switch (recordProps.type) {
+      case SXE.TYPES.ELEMENT: {
+        let { rid, name } = recordProps;
+        let view = this.viewManager.createView(name);
+        if (!view)
+          return;
+        view.rid = rid;
+        this.viewManager.register(view);
+        return view;
+      }
+      case SXE.TYPES.ATTR: {
+        let { parent, chdata, rid, name } = recordProps;
+        let view;
+        if (name === SVG.PROPERTIES.D) {
+          //path objects cannot be correctly displayed without the "command" or "d" parameter
+          //we have to re-create the path because fabricjs doesnt respond to dynamic changes to the path
+          let old = this.viewManager.getView(parent);
+          this.viewManager.deleteView(parent);
+          //create the correct path
+          view = this.viewManager.createView(SVG.PATH, chdata);
+          //copy all attributes
+          view.rid = old.rid;
+          for (let attr of Object.values(old.$attrs))
+            view.$createAttr(attr.name, attr.chdata, attr.rid);
+          this.viewManager.register(view);
+        } else {
+          view = this.viewManager.getView(parent);
+        }
+        if (!view)
+          return;
+        let attr = view.$createAttr(name, chdata, rid);
+        delete attr.isDirty;
```

Comment:
why do you have to remove this attribute `isDirty`?

Reply:
Yes, I understand the confusion, I changed it to `attr.isDirty = false;`

in JavaScript accessing an attribute that does not exist on an object will return `undefined` which behaves almost exactly the same as `false` when casted to boolean.